api/types/device.go

```diff
@@ -0,0 +1,244 @@
+	toTimePB := func(t *time.Time) *timestamppb.Timestamp {
+		if t == nil {
+			return nil
+		}
+		return timestamppb.New(*t)
+	}
+
+	osType, err := ResourceOSTypeFromString(res.Spec.OsType)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	enrollStatus, err := ResourceEnrollStatusFromString(res.Spec.EnrollStatus)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var cred *devicepb.DeviceCredential
+	if res.Spec.Credential != nil {
+		cred = &devicepb.DeviceCredential{
+			Id:           res.Spec.Credential.Id,
+			PublicKeyDer: res.Spec.Credential.PublicKeyDer,
+		}
+	}
+
+	collectedData := make([]*devicepb.DeviceCollectedData, len(res.Spec.CollectedData))
+	for i, d := range res.Spec.CollectedData {
+		dataOSType, err := ResourceOSTypeFromString(d.OsType)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		collectedData[i] = &devicepb.DeviceCollectedData{
+			CollectTime:  toTimePB(d.CollectTime),
+			RecordTime:   toTimePB(d.RecordTime),
+			OsType:       dataOSType,
+			SerialNumber: d.SerialNumber,
+		}
+	}
+
+	return &devicepb.Device{
+		ApiVersion:    res.Version,
+		Id:            res.Metadata.Name,
+		OsType:        osType,
+		AssetTag:      res.Spec.AssetTag,
+		CreateTime:    toTimePB(res.Spec.CreateTime),
+		UpdateTime:    toTimePB(res.Spec.UpdateTime),
+		EnrollStatus:  enrollStatus,
+		Credential:    cred,
+		CollectedData: collectedData,
+	}, nil
+}
+
+// DeviceToResource converts an API devicepb.Device to a resource DeviceV1 and
+// assigns all default fields.
+func DeviceToResource(dev *devicepb.Device) *DeviceV1 {
+	toTimePtr := func(pb *timestamppb.Timestamp) *time.Time {
+		if pb == nil {
+			return nil
+		}
+		t := pb.AsTime()
+		return &t
+	}
+
+	var cred *DeviceCredential
+	if dev.Credential != nil {
+		cred = &DeviceCredential{
+			Id:           dev.Credential.Id,
+			PublicKeyDer: dev.Credential.PublicKeyDer,
+		}
+	}
+
+	collectedData := make([]*DeviceCollectedData, len(dev.CollectedData))
+	for i, d := range dev.CollectedData {
+		collectedData[i] = &DeviceCollectedData{
+			CollectTime:  toTimePtr(d.CollectTime),
+			RecordTime:   toTimePtr(d.RecordTime),
+			OsType:       ResourceOSTypeToString(d.OsType),
+			SerialNumber: d.SerialNumber,
+		}
+	}
+
+	res := &DeviceV1{
+		ResourceHeader: ResourceHeader{
+			Kind:    KindDevice,
+			Version: dev.ApiVersion,
+			Metadata: Metadata{
+				Name: dev.Id,
+			},
+		},
+		Spec: &DeviceSpec{
+			OsType:        ResourceOSTypeToString(dev.OsType),
+			AssetTag:      dev.AssetTag,
+			CreateTime:    toTimePtr(dev.CreateTime),
+			UpdateTime:    toTimePtr(dev.UpdateTime),
+			EnrollStatus:  ResourceEnrollStatusToString(dev.EnrollStatus),
+			Credential:    cred,
+			CollectedData: collectedData,
+		},
+	}
+	_ = res.CheckAndSetDefaults() // assign default fields
+	return res
+}
+
+// ResourceOSTypeToString converts OSType to a string representation suitable
+// for use in resource fields.
+func ResourceOSTypeToString(osType devicepb.OSType) string {
```

Comment:
All methods from here on moved from lib/devicetrust. They could be made private now, but leaving as-is.